CHANGELOG.md

```diff
@@ -23,6 +23,10 @@ Dropping a requirement of a major version of a dependency is a new contract.
 ## [Unreleased]
 [Unreleased]: https://github.com/atlassian-labs/db-replica/compare/release-1.1.0...master
 
+### Added
+- `KeyBasedThrottledCache`
+- `SuppliedCache<T>`
```

Comment:
Why do we need it? Do we plan to use different suppliers every call? Can't we use Cache and inject the supplier via constructor?

Reply:
And if the supplier can change from call to call, how can we trust the previous supplier compatible with ours (when we read the stale data)?

Reply:
Suppliers are not taking inputs as a argument. Connection in our impl can be different on every call.

Reply:
There are 2 impls, one is for backward compatibility in a sense.

Reply:
> Suppliers are not taking inputs as a argument. Connection in our impl can be different on every call.

Yes, but You can have multiple implementations of the supplier. If there will be only one implementation then we can inject it via the constructor.

Reply:
Ok, so there's one implementation of the supplier. It's just a way to couple `AuroraPostgresLsnReplicaConsistency` and the cache, as the cache needs to call the supplier from `AuroraPostgresLsnReplicaConsistency`. `AuroraPostgresLsnReplicaConsistency` can't create the cache and inject supplier by the constructor as it's provided by a user. 🤔 


1.

Reply:
Did you have a chance to look at test cases? These should be displaying we can't inject supplier via constructor.

Reply:
> Did you have a chance to look at test cases?

Yes. I've seen it. It doesn't explain why we can't inject the values via a constructor. I think the test cases are misleading because every call uses a different supplier, so it shouldn't accept stale data generated by a different supplier.

```
 public void staleWhileInvalidating() throws InterruptedException {
        KeyBasedThrottledCache cache = new KeyBasedThrottledCache(randomKey());
        Long staleValue = 1L;
        cache.get(() -> staleValue);

        asyncCacheUpdateWithSlowSupplier(cache, 2L);

        Optional<Long> actual = cache.get(() -> 3L);

        assertThat(actual).hasValue(staleValue);
    }
```

 `Optional<Long> actual = cache.get(() -> 3L);` returns 1 for the `() -> 3L` supplier, but it shouldn't ever return a value different than 3 for this supplier. It's an logical error. The data is not stale, it's incorrect.

Reply:
The correct test should use one supplier that increases the value for each call. Then it would show the data is stale, and it would show the supplier can be injected via constructor.

Reply:
In practice, we'll have different connections and potentially different values on each supplier call.
We would need a function `(connection) -> long` as a supplier to inject is via constructor.

Reply:
I think we can decouple throttling and cache and use a Cache instead of introducing `SuppliedCache`. 🤔

Reply:
How would you implement put and reset from com.atlassian.db.replica.spi.Cache so it makes sense?

Reply:
If the supplier signature is `() -> long` and we inject it via a constructor. How to provide different connections? (every thread has a different connection, but the cache is shared across multiple threads)

Reply:
> It's an logical error.
Only if cache.get would be a pure function, it depends on the external state. Hence it's hard to reason about 🤷  .

Reply:
Yes, we work in context of `ReplicaConsistency#isConsistent(Supplier<Connection> replica)`. We need to pass the connection dynamically. I thought we might move throttling from the cache to ReplicaConsistency implementation (ThrottledConsistency), but there's no enough data in ReplicaConsistency scope. I think it needs to stay as-is.